src/lib/us-grid-emissions/apis/eia-api.ts

```diff
@@ -0,0 +1,110 @@
+        searchParams.append("api_key", apiKey);
+
+        let offset = 0;
+        let hasMoreData = true;
+        let allData: HourlyFuelTypeGeneration[] = [];
+
+        while (hasMoreData) {
+            searchParams.set("offset", String(offset));
+
+            try {
+                const url = `${BASE_URL}/${FUEL_TYPE_DATA_ROUTE}?${searchParams.toString()}`;
+                const response = await fetch(url);
+                if (!response.ok) {
+                    throw new Error(`EIA API fuel-type-data request failed: Status ${response.status}`);
+                }
+
+                const responseJson = await response.json();
+                if (responseJson.response && responseJson.response.errors) {
+                    throw new Error(`EIA API fuel-type-data request returned errors: responseJson.response.errors.join('; ')`);
+                }
+
+                allData = allData.concat(responseJson.response.data);
+                hasMoreData = responseJson.response.data.length === MAX_ROWS;
+                offset += MAX_ROWS;
+            } catch (error) {
+                throw new APIRequestError(`EIA API fuel-type-data request failed: ${error}`);
+            }
+        }
+
+        return allData;
+    };
+
+    const calculateEmissions = async (balancingAuthority: string, startDate: string, endDate: string): Promise<Record<string, number>> => {
+        const fuelTypeData = await fetchFuelTypeData(balancingAuthority, startDate, endDate);
+
+        // Group hourly generation per fuel type by period
+        const groupedByPeriod = fuelTypeData.reduce((acc: Record<string, HourlyFuelTypeGeneration[]>, current: HourlyFuelTypeGeneration) => {
+            if (!acc[current.period]) {
+                acc[current.period] = [];
+            }
+            acc[current.period].push(current);
+            return acc;
+        }, {});
+
+        console.log(groupedByPeriod);
+
+        // Calculate emissions for each period
+        const emissionsByPeriod = Object.keys(groupedByPeriod).reduce((acc: Record<string, number>, period: string) => {
+            const periodData = groupedByPeriod[period];
+            acc[period] = periodData.reduce((sum: number, data: HourlyFuelTypeGeneration) => {
+                let fuelType = data.fueltype;
+                if (fuelType === 'NG') {
+                    fuelType = 'GAS';
+                }
+
+                if (!(fuelType in CO2_EMISSIONS_FACTORS)) {
+                    throw new UnrecognizedFuelTypeError(`Unrecognized fuel type, cannot accurately calculate emissions: ${data.fueltype}. Please contact developers to add fuel type support.`)
+                }
+                const emissionsFactor = CO2_EMISSIONS_FACTORS[data.fueltype] | 0;
+                const value = parseFloat(data.value);
+                return sum + (value * emissionsFactor);
+            }, 0);
+            return acc;
+        }, {});
+
+        return emissionsByPeriod;
```

Comment:
Eventually we will need to include carbon intensity. Do you want to include that in this PR or scope additional work to make that enhancement?